### ISSUE TITLE
Kubernetes control-plane-online won't start... again

### ISSUE THREAD

**Issue by azazel75** [NixOS/nixpkgs#60687]
## Issue description
#60415 while fixing something, introduced an an error in the "pre" script: here's the fix:
```diff
diff --git a/nixos/modules/services/cluster/kubernetes/default.nix b/nixos/modules/services/cluster/kubernetes/default.nix
index 5e46bfc4240..b941d90e0ce 100644
--- a/nixos/modules/services/cluster/kubernetes/default.nix
+++ b/nixos/modules/services/cluster/kubernetes/default.nix
@@ -275,7 +275,7 @@ in {
         before = [ "kube-control-plane-online.target" ];
         path = [ pkgs.curl ];
         preStart = ''
-          until curl -Ssf ${cfg.apiserverAddress}/healthz do
+          until curl -Ssf ${cfg.apiserverAddress}/healthz; do
             echo curl -Ssf ${cfg.apiserverAddress}/healthz: exit status $?
             sleep 3
           done
```
But furthemore, after fixing this I had another problem with this -pre script: curl cannot check the apiserver's certificate (because the certificate isn't signed by any "official" CA) and I got tons of log lines like:
```
curl: (60) SSL certificate problem: unable to get issuer certificate
More details here: https://curl.haxx.se/docs/sslcerts.html
curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```
I'm using  easyCert with an externally generated certificate... i've tried  adding a `--cacert /var/lib/kubernetes/secrets/ca.pem` but still it doesn't work... So in the end I've added a `-k` option:
```diff
diff --git a/nixos/modules/services/cluster/kubernetes/default.nix b/nixos/modules/services/cluster/kubernetes/default.nix
index 5e46bfc4240..5390bc85c5c 100644
--- a/nixos/modules/services/cluster/kubernetes/default.nix
+++ b/nixos/modules/services/cluster/kubernetes/default.nix
@@ -275,8 +275,8 @@ in {
         before = [ "kube-control-plane-online.target" ];
         path = [ pkgs.curl ];
         preStart = ''
-          until curl -Ssf ${cfg.apiserverAddress}/healthz do
-            echo curl -Ssf ${cfg.apiserverAddress}/healthz: exit status $?
+          until curl -k -Ssf ${cfg.apiserverAddress}/healthz; do
+            echo curl -k -Ssf ${cfg.apiserverAddress}/healthz: exit status $?
             sleep 3
           done
         '';
```

So in the end I'm trying to understand if I'm the only one having this problem  due to my setup or, because the tests maybe didn't catch the former one, no one noticed.

### Steps to reproduce
Just enable kubernetes and do a `systemctl status`

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 5.0.8, NixOS, 19.03pre-git (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2.2`
 - channels(root): `"nixos-18.09.2526.ce0ec21d6cc"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

It's a 19.03 with the kubernetes modules from master

**Comment by azazel75** [comment]
@srhb @johanot

**Comment by srhb** [comment]
Woops, the typo must have snuck in with some last minute changes. I'll go ahead and fix that and wait for johanot on the rest.

**Comment by srhb** [comment]
@calbrecht I feel like the two `wantedBy`s in `kube-control-plane-online.{service,target}` should both be `requiredBy` -- then the test would have caught this, right?

**Comment by srhb** [comment]
Though the cross-node dependency scares me a little, and I'm not sure that the kubelets should really have such a hard dependency there... Maybe we should just wait for the control plane explicitly in the tests.

**Comment by srhb** [comment]
Maybe `-k` is the right call indeed.. It just feels fishy.

**Comment by azazel75** [comment]
@srhb yes indeed, I would be happier if curl could properly validate apiserver's cert,  but I don't know why `--cert` doesn't work in my case, It happens to you as well?

**Comment by srhb** [comment]
It works if I point --cacert to the right ca, but who's to say it exists.

**Comment by azazel75** [comment]
Digging a little more on this, I found out that while `curl --cacert /var/lib/kubernetes/secrets/ca.pem https://apiserver_fqdn:6443`  doesn't validate the server certificate and outputs the complain I posted before, trying with `wget --ca-certificate=/var/lib/kubernetes/secrets/ca.pem https://apiserver_fqdn:6443` validates it, so I'm trying to figure out if there's the chance that maybe my CA certificates is missing some flags that curl cares about but wget doesn't

**Comment by johanot** [comment]
@srhb I feel that the insecure flag would be an ok and pragmatic solution. However, we should be able to set `CURL_CA_BUNDLE=${cfg.caFile or ""}` or similar. We can expect users that do custom PKI to manually set `services.kubernetes.caFile`. However, I think the most worrisome here is that the kubernetes test does not capture the error - as you mention. We have to sort of out the dependencies such that a failing control-plane-online service/target doesn't look like a success.

**Comment by saschagrunert** [comment]
Hey, I'm really looking forward tot his fix, since unstable is currently broken because of this issue. 😕

**Comment by azazel75** [comment]
@saschagrunert put this in your module:
```
# Temporary off-nixpkgs tree fix, see
# https://github.com/NixOS/nixpkgs/issues/60687
systemd.services.kube-control-plane-online.preStart =
  let
    cfg = config.services.kubernetes;
  in lib.mkForce ''
    until curl -k -Ssf ${cfg.apiserverAddress}/healthz; do
      echo curl -k -Ssf ${cfg.apiserverAddress}/healthz: exit status $?
      sleep 3
    done
'';
```

**Comment by peteryuu** [comment]
Even with this fix (much appreciated), there seems to be a syntax error in the script. I have no idea why.

`May 26 00:20:33 nixos 0lbwlqirr9mq539vj267dvrr7ffzsv97-unit-script-kube-control-plane-online-pre-start[15663]: /nix/store/0lbwlqirr9mq539vj267dvrr7ffzsv97-unit-script-kube-control-plane-online-pre-start: line 5: syntax error near unexpected token `done'`

Here's the script contents:
```sh
#! /nix/store/x0mlaj4z4ciycaycfwc36l1932mwywfj-bash-4.4-p23/bin/bash -e
until curl -k -Ssf https://127.0.0.1:6443/healthz; do
echo curl -k -Ssf https://127.0.0.1:6443/healthz: exit status $?
sleep 3
done
```

**Comment by azazel75** [comment]
Mine works flawlessly, but I don't  see the indentation, not that this fragment seems wrong...

**Comment by peteryuu** [comment]
My bad. The problem was with my text editor, which wrote CRLF endings and threw bash out of whack. It works perfectly now, thanks!

**Comment by johanot** [comment]
I thiink this can be closed as well, following the merge of #67563 ? @azazel75 @srhb Am I right? Anyone interested in developing on the kubernetes module, please have your say here: https://discourse.nixos.org/t/kubernetes-the-nixos-module-of-the-future/3922 :)

**Comment by azazel75** [comment]
@johanot yes, this can be closed
